wasmtime/_memory.py

```diff
@@ -62,6 +62,73 @@ def data_ptr(self, store: Storelike) -> "ctypes._Pointer[c_ubyte]":
         """
         return ffi.wasmtime_memory_data(store._context, byref(self._memory))
 
+    def set_store(self, store: Storelike):
+        """
+        must be called to set the store for highlevel access of memory
+        """
+        self._store=store
+
+    def __getitem__(self, key: int):
+        """
+        provide memory[offset] or memory[start:stop]
+        """
+        if self._store is None:
+            raise RuntimeError("you must call `set_store()` before using highlevel access")
+        data_ptr = self.data_ptr(self._store)
+        size = self.data_len(self._store)
+        if isinstance(key, int):
+            if key>=size:
+                raise IndexError("out of memory size")
+            return data_ptr[key]
+        if not isinstance(key, slice):
+            raise TypeError("key can either be integer index or slice")
+        
+        start, stop, step = key.indices(size)
+        if step!=1:
+            raise ValueError("slice with step is not supported")
+
+        val_size = stop - start
+        value = bytearray(val_size)
+
+        if stop>size:
+            raise IndexError("out of memory size")
+        dst_ptr = (ctypes.c_ubyte * val_size).from_buffer(value)
+        src_ptr = ctypes.addressof((ctypes.c_ubyte*val_size).from_address(ctypes.addressof(data_ptr.contents)+start))
+        ctypes.memmove(dst_ptr, src_ptr, val_size)
+        return value
+
+    def __setitem__(self, key, value: bytearray):
+        """
+        provide setter for memory[key] with slice support
+
+        memory[offset]=value
+        memory[start:stop]=b'hello world'
+        memory[start:stop]=bytearray([1,2,3])
+        """
+        if self._store is None:
+            raise RuntimeError("you must call `set_store()` before using highlevel access")
+        data_ptr = self.data_ptr(self._store)
+        size = self.data_len(self._store)
+        if isinstance(key, int):
+            if key>=size:
+                raise IndexError("out of memory size")
+            data_ptr[key] = value
+        if not isinstance(key, slice):
+            raise TypeError("key can either be integer index or slice")
+
+        start, stop, step = key.indices(size)
+        if step!=1:
+            raise ValueError("slice with step is not supported")
+
+        val_size = len(value)
+        if stop is None:
+            stop=start+val_size
+        if stop-start>val_size or stop>size:
+            raise IndexError("out of memory size")
+        src_ptr = (ctypes.c_ubyte * val_size).from_buffer(value)
+        dst_ptr = ctypes.addressof((ctypes.c_ubyte*val_size).from_address(ctypes.addressof(data_ptr.contents)+start))
```

Comment:
The `ctypes.c_ubyte * val_size` type is "calculated" twice here, could that be refactored to happen only once?

Reply:
> The `ctypes.c_ubyte * val_size` type is "calculated" twice here, could that be refactored to happen only once?

I have an impression this is just a no-cost type casting, but let me review this and get back to you

```python
    #pr=Profile()
    #pr.enable()
    #for i in range(10000):
    #    cdb_djp_hash_wasm(large_a)
    #pr.disable()
    #pr.print_stats()
    #pr.print_stats('cumulative')
    #pr.print_stats('calls')
```

Reply:
I made it once now